packages/rum-core/src/domain/contexts/pageStateContexts.ts

```diff
@@ -0,0 +1,157 @@
+
+export const enum PageState {
+  ACTIVE = 'active',
+  PASSIVE = 'passive',
+  HIDDEN = 'hidden',
+  FROZEN = 'frozen',
+  TERMINATED = 'terminated',
+}
+
+type PageStateTimeline = Array<{ state: PageState; startTime: RelativeTime }>
+
+type Counter = { count: number; duration: Duration }
+export type PageStateContext = {
+  active: Counter
+  passive: Counter
+  hidden: Counter
+  frozen: Counter
+  terminated: Counter
+}
+
+export interface PageStateContexts {
+  getPageStates: (startTime: RelativeTime, duration: Duration) => PageStateContext | undefined
+  stop: () => void
+}
+
+let pageStateTimeline: PageStateTimeline = []
+let state: PageState | undefined
+
+export function startPageStateContexts(): PageStateContexts {
+  if (!isExperimentalFeatureEnabled('resource_page_states')) {
+    return {
+      getPageStates: () => undefined,
+      stop: noop,
+    }
+  }
+
+  addPageState(getState())
+
+  const listenerOpts = { capture: true }
+  const stoppers: Array<{ stop: () => void }> = []
+
+  stoppers.push(
+    addEventListeners(
+      window,
+      [DOM_EVENT.PAGE_SHOW, DOM_EVENT.FOCUS, DOM_EVENT.BLUR, DOM_EVENT.VISIBILITY_CHANGE, DOM_EVENT.RESUME],
+      (ev) => {
+        if (ev.isTrusted) addPageState(getState())
+      },
+      listenerOpts
+    ),
+    addEventListener(
+      window,
+      DOM_EVENT.FREEZE,
+      (ev) => {
+        if (ev.isTrusted) addPageState(PageState.FROZEN)
+      },
+      listenerOpts
+    ),
+    addEventListener(
+      window,
+      DOM_EVENT.PAGE_HIDE,
+      (ev: PageTransitionEvent) => {
+        // If the event's persisted property is `true` the page is about
+        // to enter the back/forward cache, which is also in the frozen state.
+        // If the event's persisted property is not `true` the page is
+        // about to be unloaded.
+        if (ev.isTrusted) addPageState(ev.persisted ? PageState.FROZEN : PageState.TERMINATED)
+      },
+      listenerOpts
+    )
+  )
+
+  function findStartIndex(startTime: RelativeTime) {
+    let i = -1
+    let endTime
+
+    do {
+      i++
+      endTime = i + 1 < pageStateTimeline.length ? pageStateTimeline[i + 1].startTime : Infinity
+    } while (i < pageStateTimeline.length && endTime <= startTime)
+
+    return i
+  }
+
+  return {
+    getPageStates(startTime: RelativeTime, duration: Duration) {
+      const endTime = addDuration(startTime, duration)
+      if (pageStateTimeline.length === 0 || pageStateTimeline[0].startTime >= endTime) return
+
+      const pageStateContext: PageStateContext = {
+        active: { count: 0, duration: 0 as Duration },
+        passive: { count: 0, duration: 0 as Duration },
+        hidden: { count: 0, duration: 0 as Duration },
+        frozen: { count: 0, duration: 0 as Duration },
+        terminated: { count: 0, duration: 0 as Duration },
+      }
+
+      let i = findStartIndex(startTime)
+      while (i < pageStateTimeline.length && pageStateTimeline[i].startTime < endTime) {
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**🥜 nitpick:** ‏a `for(let i = 0; i < ... && ...; i++)` loop might be more idiomatic